contribute/architecture/k8s-inspired-backend-arch.md

```diff
@@ -0,0 +1,288 @@
+
+On top of adopting the declarative, eventually-consistent system design, we are also adopting Kubernetes API's strict approach to URL structures, versioning and schemas. This approach addresses limitations of Grafana's traditional Legacy APIs (`/api/...`), including:
+- **Inconsistent designs:** Variations in structure, parameters, and response formats across different endpoints.
+- **No versioning:** Difficulty managing breaking changes and API evolution without clear versioning.
+- **Absence of schemas:** Lack of machine-readable schemas hindered programmatic interaction, validation, and tooling.
+- **Limited extensibility model:** No standardized way for Grafana plugins to extend core APIs.
+
+The end goal is for the Resource APIs to become the only interface for managing core Grafana resources. We will eventually deprecate and remove all Legacy API endpoints.
+
+## 1. Key Differences from Legacy API Endpoints
+
+- **URL structure & versioning:**
+  - **Resource APIs:** Follow Kubernetes conventions (`/apis/<group>/<version>/namespaces/<namespace>/<resource>/<name>`). Includes explicit API versions (e.g., `v0alpha1`, `v1alpha1`, `v2alpha1`) in the path, allowing for controlled evolution and multiple versions of a resource API to co-exist.
+  - **Legacy APIs:** Variable path structures (e.g., `/api/dashboards/uid/:uid`, `/api/ruler/grafana/api/v1/rules/:uid/:uid`). Less explicit versioning.
+
+- **Resource schemas:**
+  - **Resource APIs:** Each resource has a well-defined schema (`spec`) and is wrapped with an envelope with common metadata, all APIs come with an always-in-sync OpenAPI spec.
+  - **Legacy APIs:** Structures vary
+
+- **Namespacing / org context:**
+  - **Resource APIs:** Uses explicit namespaces in the path (`/namespaces/<namespace>/...`) mapping to Grafana Organization IDs for scoping.
+  - **Legacy APIs:** Org context determined implicitly (session, API key), not usually part of the URL structure.
+
+- **Consistency of convenience features:**
+  - **Resource APIs:** Convenience features such as resource history or restore come out of the box; a single implementation works across all resources
+  - **Legacy APIs:** Different APIs support different convenience features; implementations are feature-specific
+
+- **Error Handling:**
+  - **Resource APIs:** Uses Kubernetes error response conventions (HTTP status codes 400, 401/403, 404, 409; structured error body).
+  - **Legacy APIs:** Variable error formats; Grafana-specific.
+
+## 2. Core Concepts
+
+- **Resource APIs:** New Grafana API endpoints that mimic the Kubernetes API structure (e.g., `/apis/dashboard.grafana.app/v1alpha1/dashboards`). These are served by Grafana itself and are distinct from any Kubernetes API running on the same system..
+
+- **API implementation approaches:**
+  - **Registry Approach (`pkg/registry/apis/...`):** Original method using Go code (`.../register.go` files)
```

Comment:
We still have relevant code in registry/apis even if the schemas are defined in apps (such as legacy fallbacks)

Reply:
true - added a note